client/transaction-pool/src/lib.rs

```diff
@@ -727,6 +774,220 @@ where
 	}
 }
 
+/// Trait for deciding if core part of maintenance procedure shall be executed.
+///
+/// For the following chain:
+///   B1-C1-D1-E1
+///  /
+/// A
+///  \
+///   B2-C2-D2-E2
+///
+/// Some scenarios and expected behavior:
+/// nbb(C1), f(C1) -> false (handle_enactmanet was already performed in nbb(C1))
+/// f(C1), nbb(C1) -> false (handle_enactmanet was already performed in f(C1))
+///
+/// f(C1), nbb(D2) -> false (handle_enactmanet was already performed in f(C1), we should not retract
+/// finalized block) f(C1), f(C2), nbb(C1) -> false
+///
+/// nbb(C1), nbb(C2) -> true (switching fork is OK)
+pub trait EnactmentPolicy<Block: BlockT> {
+	/// Based on provided ChainEvent the decision if maintainance/enactment procedure for evented
+	/// header shall proceed.
+	/// Additionally enacted and retracted paths are returned respectively.
+	fn should_handle_enactment(
+		&self,
+		event: &ChainEvent<Block>,
+	) -> (bool, Arc<Vec<Block::Hash>>, Arc<Vec<Block::Hash>>);
+}
+
+struct EnactmentHelper<Block, PoolApi>
+where
+	Block: BlockT,
+	PoolApi: graph::ChainApi<Block = Block>,
+{
+	best_block: Arc<Mutex<Option<Block::Hash>>>,
+	finalized_block: Arc<Mutex<Option<Block::Hash>>>,
+	metadata_cache: Arc<Mutex<HeaderMetadataCache<Block>>>,
+	api: Arc<PoolApi>,
+}
+
+impl<Block, PoolApi> EnactmentPolicy<Block> for EnactmentHelper<Block, PoolApi>
+where
+	Block: BlockT,
+	PoolApi: 'static + graph::ChainApi<Block = Block>,
+{
+	fn should_handle_enactment(
+		&self,
+		event: &ChainEvent<Block>,
+	) -> (bool, Arc<Vec<Block::Hash>>, Arc<Vec<Block::Hash>>) {
+		match event {
+			ChainEvent::NewBestBlock { hash, tree_route } => {
+				println!("--------------------------------------------------------------------------------");
+				println!("process_new_best_block_event");
+				let (enacted, retracted) = if let Some(tree_route) = tree_route {
+					(
+						tree_route.enacted().iter().map(|block| block.hash).collect(),
+						tree_route.retracted().iter().map(|block| block.hash).collect(),
+					)
+				} else {
+					(vec![], vec![])
+				};
+
+				let result = self.resolve(*hash, &enacted, &retracted, false);
+
+				(result, Arc::from(enacted), Arc::from(retracted))
+			},
+			ChainEvent::Finalized { hash, tree_route } => {
+				println!("--------------------------------------------------------------------------------");
+				println!("process_finalized_event");
+				let tree_route = Arc::new(tree_route.to_vec()); //todo: cleanup
+				let result = self.resolve(*hash, &tree_route, &vec![], true);
+				(result, tree_route, Arc::from(vec![]))
+			},
+		}
+	}
+}
+
+impl<Block, PoolApi> EnactmentHelper<Block, PoolApi>
+where
+	Block: BlockT,
+	PoolApi: 'static + graph::ChainApi<Block = Block>,
+{
+	fn new(api: Arc<PoolApi>) -> Self {
+		EnactmentHelper {
+			best_block: Arc::new(Mutex::new(None)),
+			finalized_block: Arc::new(Mutex::new(None)),
+			metadata_cache: Arc::new(Mutex::new(HeaderMetadataCache::new(16))),
```

Comment:
Just add the `tree_route` function to the `api`. No need to have an extra cache etc here.